Wallet quickstart (#1447)
TBD54566975/developer.tbd.website: site/docs/tbdex/quickstart.mdx

---
sidebar_position: 1
title: Wallet Quickstart
hide_title: true
hide_table_of_contents: false
---

import { QuickstartExecutionProvider } from '@site/src/components/QuickstartExecutionContext';
import Version from '@site/src/components/Version';
import PackageJson from '@site/src/components/PackageJson';
import CodeSnippet from '@site/src/components/CodeSnippet';
import WalletQuickstartCodeRunner from '@site/src/components/WalletQuickstartCodeRunner';

<QuickstartExecutionProvider>
# tbDEX Wallet in 5️⃣ minutes

## Overview

The role of a Wallet App in a tbDEX ecosystem is to find liquidity for its customers. Let’s build a tbDEX-enabled wallet in under 5 minutes. You will learn how to:

✅  Integrate the tbDEX SDK into a wallet app

✅  Connect a wallet app with a PFI (Participating Financial Institution)

✅  Use a wallet app to send tbDEX messages to a PFI to facilitate an exchange

Let’s go!

## Wallet App Basics

In this exercise, your app is responsible for helping your customer, Alice, exchange `USD` for `KES`. First, we'll create a [DID (Decentralized Identifier)](/docs/web5/learn/decentralized-identifiers) for Alice.

### Create Customer's DID

> In order for you to communicate with PFIs on Alice's behalf, she'll need a DID (as will all of your customers). Here's how to create a DID:

Reviewer: ```suggestion
In order for you to communicate with PFIs on your customers' behalf, they will need a DID. Let's create one for Alice now:
```

<CodeSnippet snippetName="walletQuickstartDidCreate" />

<WalletQuickstartCodeRunner executeCodeName="executeDidCreate" stepIndex={0} customPlaceholder={"See your DID"}/>

## Connect to a PFI

> To connect to a PFI, all you need is their DID. Let's get connected to a mock PFI whose DID is `did:dht:ccqamm6qgbe763ya8f3bo5mkrtxx1pz7i789zeqq4bmoae4qnixy`.

Reviewer: this section seems a bit off because there's no action. then we use a variable named `pfiDid` in the next section.

You should probably add a code block here

```js
const pfiDid = did:dht:ccqamm6qgbe763ya8f3bo5mkrtxx1pz7i789zeqq4bmoae4qnixy
```

Author: Added

> To connect to a PFI, all you need is their DID. Let's get connected to a mock PFI whose DID is `did:dht:ccqamm6qgbe763ya8f3bo5mkrtxx1pz7i789zeqq4bmoae4qnixy`.

Reviewer: ```suggestion
To connect to a PFI, all you need is their DID. Let's use the DID of a mock PFI: 
```

## Get Offerings

> [Offerings](/docs/tbdex/message-types#offerings) describe the currency pair and transaction parameters. With the PFI's DID, your app can fetch all of their offerings:

Reviewer: ```suggestion
All PFIs have [Offerings](/docs/tbdex/message-types#offerings) which are a menu of services they provide. Offerings describe the currency pairs and transaction parameters offered . Let's fetch all of the offerings from the PFI:
```

<CodeSnippet snippetName="walletQuickstartGetOfferings"/>

<WalletQuickstartCodeRunner executeCodeName="executeGetOfferings" stepIndex={1} autoOpenDetails={false} customPlaceholder={"See the PFI's Offerings"} viewJsonObj={true} collapseLevel={1}/>

Once you have all of the PFI's offerings, you can filter it based on a variety of factors, including the currency pair you're looking for. In this case, you're looking for an offering that exchanges `USD` for `KES`:

<CodeSnippet snippetName="walletQuickstartSelectOffering" />

#### Obtain Verifiable Credentials

> The Offering you received will likely have a `requiredClaims` section that specifies which, if any, [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) (VCs) you’ll need to present in order to be eligible for the Offering. Begin by applying for your VC:

Reviewer: ```suggestion
The Offering we selected has a `requiredClaims` section that specifies [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) (VCs) Alice needs  in order to be eligible for the Offering. 
```

show a code block with the requiredClaims part of the offering.

Then add:

> The PFI requires Alice to have a Sanctions credential. Let's help Alice apply for this VC.

Reviewer: Chris, see the other part of the comment:

show a code block with the requiredClaims part of the offering.

Then add:

> The PFI requires Alice to have a Sanctions credential. Let's help Alice apply for this VC.

right now. it's still a bit confusing without this context

![Screenshot_20240712_095527_Chrome](https://github.com/user-attachments/assets/1d83b8f8-f630-410e-bba8-bdcc683baad2)

Author: Clarified, added a snippet of the `requiredClaims` as well as more explicit `SanctionsCredential` application text

> <WalletQuickstartCodeRunner executeCodeName="executeApplyForCredential" stepIndex={2} customPlaceholder={"Apply for a SanctionCredential"}/>

Reviewer: should show an assignment to `myCredentials`. this variable is used in the next step but hasnt been introduced

Author: Totally understand where you're coming from wanting to show assignment of `myCredentials`, and I looked for how else we showed the usage of the `PresentationExchange.selectCredentials()` method in our docs ([found here](https://developer.tbd.website/docs/tbdex/wallet/managing-credentials#obtain-eligible-offerings)). In the other example we're just kicking the problem up one layer where we hard-code in JWT values, which is potentially more confusing(?) My thought is that it's sufficient to describe in text outside the code block what `myCredentials` is instead of adding in more code that doesn't really make it all that much clearer

Author: @angiejones see updated text for explainer

Reviewer: ok, gonna rephrase it a bit

Author: Resolving after applied suggestion

```

## Get Offerings

All PFIs have [Offerings](/docs/tbdex/message-types#offerings) which are a menu of services they provide. Offerings describe the currency pairs and transaction parameters offered . Let's fetch all of the offerings from the PFI:

<CodeSnippet snippetName="walletQuickstartGetOfferings"/>

<WalletQuickstartCodeRunner executeCodeName="executeGetOfferings" stepIndex={1} autoOpenDetails={false} customPlaceholder={"See the PFI's Offerings"} viewJsonObj={true} collapseLevel={1}/>

Once you have all of the PFI's offerings, you can filter it based on a variety of factors, including the currency pair you're looking for. In this case, you're looking for an offering that exchanges `USD` for `KES`:

<CodeSnippet snippetName="walletQuickstartSelectOffering" />

#### Obtain Verifiable Credentials

The Offering we selected has a `requiredClaims` section that specifies [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) (VCs) Alice needs  in order to be eligible for the Offering. 

<WalletQuickstartCodeRunner executeCodeName="executeApplyForCredential" stepIndex={2} customPlaceholder={"Apply for a SanctionCredential"}/>

> VCs are stored as JWTs - in this case we're assuming that they have been fetched from local storage and placed into a string array called `myCredentials` -, and once you've saved the token above you can access and filter through them like so:

Reviewer: ```suggestion
VCs are stored as JSON Web Tokens (JWTs). Your wallet app has saved Alice's new VC into an array named `myCredentials`. Let's now find which of Alice’s credentials meet the PFI's requirements:
```

<CodeSnippet snippetName="walletQuickstartSelectCredentials" />

## Request a Quote

Now that you've found an offering that matches Alice's request, your app can request a quote from the PFI to obtain a formal offer based on Alice's preferred payment methods.

<CodeSnippet snippetName="walletQuickstartCreateRfq" />

<WalletQuickstartCodeRunner executeCodeName="executeCreateRfq" stepIndex={3} customPlaceholder={"See your RFQ"} viewJsonObj={true} collapseLevel={1}/>

After creating an RFQ message, you can verify that it is properly constructed and its data satisfies the Offering's requirements.

If no exceptions are thrown, you'll sign the RFQ with Alice's DID to ensure its authenticity.

Then finally, you'll send the RFQ to the PFI to initiate the transaction.

> <CodeSnippet snippetName="walletQuickstartSendRfq" />

Reviewer: Going through it, it just feels like something is missing here.

can we make this runnable where the user gets a 200 code or something? You can even fake it where they press Run and it just writes the status.

Author: added

## Process Quote

In response to your RFQ, the PFI will write a Quote message to the exchange if they are able to fulfill Alice's request, or a Close message if not. Here's how to obtain that message:

<CodeSnippet snippetName="walletQuickstartProcessQuote" />

<WalletQuickstartCodeRunner executeCodeName="executeProcessQuote" stepIndex={4} customPlaceholder={"View Quote"} viewJsonObj={true} collapseLevel={1}/> 

## Place Order

> The Quote message will contain the exact exchange rate including any fees. Your app can present this to Alice to determine if she wants to proceed with the transaction. If she does, you'll create an Order message, sign it with Alice's DID, and send it to the PFI:

Reviewer: ```suggestion
The Quote contains the exact exchange rate including any fees. Your app can present this to Alice to determine if she wants to proceed with the transaction. If she does, you'll create an Order message, sign it with Alice's DID, and send it to the PFI:
```

> <CodeSnippet snippetName="walletQuickstartCreateOrder" />

Reviewer: the sign and send are not shown

Author: added

<WalletQuickstartCodeRunner executeCodeName="executeCreateOrder" stepIndex={5} customPlaceholder={"Create your order"} viewJsonObj={true} collapseLevel={1}/> 

## End Transaction

To end the transaction, the PFI will write a Close message indicating whether the transaction successfully completed or not. They'll also include a reason for closing the transaction which will contain more information that you may find helpful in event of a failure.

> <CodeSnippet snippetName="walletQuickstartProcessClose" />

Reviewer: show how to get `success`

Author: added

<WalletQuickstartCodeRunner executeCodeName="executeProcessClose" stepIndex={6} customPlaceholder={"View your close reason"}/> 

## Next Steps with tbDEX

> Congratulations on creating your first Wallet! In just a few minutes you were able to contact a PFI for its offerings, request a quote, place an order, and finish your transaction.

Reviewer: ```suggestion
Congratulations on creating your first Wallet! In just a few minutes you were able to contact a PFI for its offerings, request a quote, and place an order for Alice to swap USD for KES.
```

---
sidebar_position: 1
title: Wallet Quickstart
hide_title: true
hide_table_of_contents: false
---

import { QuickstartExecutionProvider } from '@site/src/components/QuickstartExecutionContext';
import Version from '@site/src/components/Version';
import PackageJson from '@site/src/components/PackageJson';
import CodeSnippet from '@site/src/components/CodeSnippet';
import WalletQuickstartCodeRunner from '@site/src/components/WalletQuickstartCodeRunner';

<QuickstartExecutionProvider>
# tbDEX Wallet in 5️⃣ minutes

## Overview

The role of a Wallet App in a tbDEX ecosystem is to find liquidity for its customers. Let’s build a tbDEX-enabled wallet in under 5 minutes. You will learn how to:

✅  Integrate the tbDEX SDK into a wallet app

✅  Connect a wallet app with a PFI (Participating Financial Institution)

> ✅  Use a wallet app to send tbDEX messages to a PFI to facilitate an exchange

Reviewer: ```suggestion
✅  Use a wallet app to send tbDEX messages to a PFI to facilitate a transaction
```

Let’s go!

## Wallet App Basics

In this exercise, your app is responsible for helping your customer, Alice, exchange `USD` for `KES`. First, we'll create a [DID (Decentralized Identifier)](/docs/web5/learn/decentralized-identifiers) for Alice.

### Create Customer's DID

In order for you to communicate with PFIs on your customers' behalf, they will need a DID. Let's create one for Alice now:

<CodeSnippet snippetName="walletQuickstartDidCreate" />

> <WalletQuickstartCodeRunner executeCodeName="executeDidCreate" stepIndex={0} customPlaceholder={"See your DID"}/>

Reviewer: ```suggestion
<WalletQuickstartCodeRunner executeCodeName="executeDidCreate" stepIndex={0} customPlaceholder={"See Alice's DID"}/>
```

## Connect to a PFI

To connect to a PFI, all you need is their DID. Let's use the DID of a mock PFI: 

```js
const pfiDid = did:dht:ccqamm6qgbe763ya8f3bo5mkrtxx1pz7i789zeqq4bmoae4qnixy
```

## Get Offerings

> All PFIs have [Offerings](/docs/tbdex/message-types#offerings) which are a menu of services they provide. Offerings describe the currency pairs and transaction parameters offered . Let's fetch all of the offerings from the PFI:

Reviewer: ```suggestion
All PFIs have [Offerings](/docs/tbdex/message-types#offerings) which are a menu of services they provide. Offerings describe the currency pairs and transaction parameters offered. Let's fetch all of the offerings from the PFI:
```

Once you have all of the PFI's offerings, you can filter it based on a variety of factors, including the currency pair you're looking for. In this case, you're looking for an offering that exchanges `USD` for `KES`:

<CodeSnippet snippetName="walletQuickstartSelectOffering" />

#### Obtain Verifiable Credentials

The Offering we selected has a `requiredClaims` section that specifies [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) (VCs) Alice needs  in order to be eligible for the Offering. This PFI requires Alice to have a `SanctionsCredential`, as outlined in the `requiredClaims` section of the Offering:

```js
{
    'id': '6bdaec32-e511-489d-b2fd-27a051bd819b',
    'input_descriptors': [
      {
        'id': '65752b5a-a454-4605-85de-7f159b22ce01',
        'purpose': 'We need to know if you are a sanctioned individual',
        'constraints': {
          'fields': [
            {
              'path': [ '$.type[*]' ],
              'filter': {
                'type': 'string',

>                 'const': 'SanctionsCredential'

Reviewer: ```suggestion
                //highlight-start
                'const': 'SanctionsCredential'
                //highlight-end
```

              }
            }
          ]
        }
      }
    ]
}
```

As a result, we'll help Alice apply for the `SanctionsCredential` she needs to satisfy the PFI's requirements:

<WalletQuickstartCodeRunner executeCodeName="executeApplyForCredential" stepIndex={2} customPlaceholder={"Apply for a SanctionCredential"}/>

> VCs are stored as JSON Web Tokens (JWTs). Your wallet app has saved Alice's new VC into an array named `myCredentials`. Let's now find which of Alice’s credentials meet the PFI's requirements:

Reviewer: ```suggestion
VCs are stored as JSON Web Tokens (JWTs). Your wallet app has saved Alice's new VC into an array named `myCredentials`. Let's now select Alice’s credential to meet the PFI's requirements:
```

<CodeSnippet snippetName="walletQuickstartSelectCredentials" />

> ## Request a Quote

Reviewer: ```suggestion
## Request Quote
```

Now that you've found an offering that matches Alice's request, your app can request a quote from the PFI to obtain a formal offer based on Alice's preferred payment methods.

<CodeSnippet snippetName="walletQuickstartCreateRfq" />

> <WalletQuickstartCodeRunner executeCodeName="executeCreateRfq" stepIndex={3} customPlaceholder={"See your RFQ"} viewJsonObj={true} collapseLevel={1}/>

Reviewer: ```suggestion
<WalletQuickstartCodeRunner executeCodeName="executeCreateRfq" stepIndex={3} customPlaceholder={"See the RFQ"} viewJsonObj={true} collapseLevel={1}/>
```

After creating an RFQ message, you can verify that it is properly constructed and its data satisfies the Offering's requirements.

If no exceptions are thrown, you'll sign the RFQ with Alice's DID to ensure its authenticity.

Then finally, you'll send the RFQ to the PFI to initiate the transaction.

<CodeSnippet snippetName="walletQuickstartSendRfq" />

<WalletQuickstartCodeRunner executeCodeName="executeSendRfq" stepIndex={4} customPlaceholder={"Submit RFQ"} viewJsonObj={true} collapseLevel={1}/> 

## Process Quote

In response to your RFQ, the PFI will write a Quote message to the exchange if they are able to fulfill Alice's request, or a Close message if not. Here's how to obtain that message:

<CodeSnippet snippetName="walletQuickstartProcessQuote" />

> <WalletQuickstartCodeRunner executeCodeName="executeProcessQuote" stepIndex={5} customPlaceholder={"View Quote"} viewJsonObj={true} collapseLevel={1}/> 

Reviewer: I get an error here

<img width="326" alt="image" src="https://github.com/user-attachments/assets/9e7a6eeb-c6fb-4ed8-9b4e-996e4dbc32af">

## Place Order

The Quote contains the exact exchange rate including any fees. Your app can present this to Alice to determine if she wants to proceed with the transaction. If she does, you'll create an Order message:

<CodeSnippet snippetName="walletQuickstartCreateOrder" />

> <WalletQuickstartCodeRunner executeCodeName="executeCreateOrder" stepIndex={6} customPlaceholder={"Create your order"} viewJsonObj={true} collapseLevel={1}/> 

Reviewer: ```suggestion
<WalletQuickstartCodeRunner executeCodeName="executeCreateOrder" stepIndex={6} customPlaceholder={"Create Alice's Order"} viewJsonObj={true} collapseLevel={1}/> 
```

Once you have an order message, you can sign it with Alice's DID and send it to the PFI to execute the transaction.

<CodeSnippet snippetName="walletQuickstartSubmitOrder" />

> <WalletQuickstartCodeRunner executeCodeName="executeSendOrder" stepIndex={7} customPlaceholder={"Submit your order"} viewJsonObj={true} collapseLevel={1}/> 

Reviewer: ```suggestion
<WalletQuickstartCodeRunner executeCodeName="executeSendOrder" stepIndex={7} customPlaceholder={"Submit Order to PFI"} viewJsonObj={true} collapseLevel={1}/> 
```

```

## Get Offerings

All PFIs have [Offerings](/docs/tbdex/message-types#offerings) which are a menu of services they provide. Offerings describe the currency pairs and transaction parameters offered. Let's fetch all of the offerings from the PFI:

<CodeSnippet snippetName="walletQuickstartGetOfferings"/>

<WalletQuickstartCodeRunner executeCodeName="executeGetOfferings" stepIndex={1} autoOpenDetails={false} customPlaceholder={"See the PFI's Offerings"} viewJsonObj={true} collapseLevel={1}/>

Once you have all of the PFI's offerings, you can filter it based on a variety of factors, including the currency pair you're looking for. In this case, you're looking for an offering that exchanges `USD` for `KES`:

<CodeSnippet snippetName="walletQuickstartSelectOffering" />

#### Obtain Verifiable Credentials

The Offering we selected has a `requiredClaims` section that specifies [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) (VCs) Alice needs  in order to be eligible for the Offering. This PFI requires Alice to have a `SanctionsCredential`, as outlined in the `requiredClaims` section of the Offering:

> ```js

Reviewer: this is different than what the selected offering has. It has 2 input descriptors

Author: Added the second input descriptor
